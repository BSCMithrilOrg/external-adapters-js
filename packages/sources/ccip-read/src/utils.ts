import { Fragment, Interface, JsonFragment } from '@ethersproject/abi'

export const toInterface = (
  abi: string | readonly (string | Fragment | JsonFragment)[] | Interface,
): Interface => {
  return Interface.isInterface(abi) ? abi : new Interface(abi)
}
