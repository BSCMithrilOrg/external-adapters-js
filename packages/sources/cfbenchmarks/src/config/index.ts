import { Requester, util } from '@chainlink/ea-bootstrap'
import { Config as BaseConfig } from '@chainlink/types'

export const NAME = 'CFBENCHMARKS'

export const ENV_API_USERNAME = 'API_USERNAME'
export const ENV_API_PASSWORD = 'API_PASSWORD'
export const ENV_API_SECONDARY = 'API_SECONDARY'

export const DEFAULT_ENDPOINT = 'values'
export const DEFAULT_API_ENDPOINT = 'https://www.cfbenchmarks.com/api'
export const DEFAULT_WS_API_ENDPOINT = 'wss://www.cfbenchmarks.com/ws/v4'
export const SECONDARY_API_ENDPOINT = 'https://unregprod.cfbenchmarks.com/api'
export const SECONDARY_WS_API_ENDPOINT = 'wss://unregprod.cfbenchmarks.com/ws/v4'

export type Config = BaseConfig & {
  useSecondary: boolean
  username: string
  password: string
}

export const makeConfig = (prefix?: string): Config => {
  const config = {
    ...Requester.getDefaultConfig(prefix),
    useSecondary: util.parseBool(util.getEnv(ENV_API_SECONDARY)),
    username: util.getRequiredEnv(ENV_API_USERNAME, prefix),
    password: util.getRequiredEnv(ENV_API_PASSWORD, prefix),
  }

  config.api.baseURL =
    config.api.baseURL || (config.useSecondary ? SECONDARY_API_ENDPOINT : DEFAULT_API_ENDPOINT)
  config.api.baseWsURL =
    config.api.baseWsURL ||
    (config.useSecondary ? SECONDARY_WS_API_ENDPOINT : DEFAULT_WS_API_ENDPOINT)

  config.defaultEndpoint = DEFAULT_ENDPOINT
  return config
}
