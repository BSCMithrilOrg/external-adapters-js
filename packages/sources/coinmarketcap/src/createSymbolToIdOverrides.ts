import * as process from 'process'
import * as fs from 'fs'

import axios from 'axios'

type OverrideObj = {
  [adapterName: string]: {
    [symbol: string]: string
  }
}

type CoinData = {
  id: number
  name: string
  symbol: string
  slug: string
  rank: number
  [otherData: string]: unknown
}

export const getSymbolToIdOverrides = async (): Promise<OverrideObj> => {
  if (!process.env.API_KEY) throw Error('API_KEY environment variable not set.')
  const headers = { 'X-CMC_PRO_API_KEY': process.env.API_KEY }
  const response = await axios.get('https://pro-api.coinmarketcap.com/v1/cryptocurrency/map', {
    headers,
  })
  const coinsList = response.data.data
  const adapterSymToIdOverrides = {} as { [symbol: string]: string }

  coinsList.forEach((coinData: CoinData) => {
    adapterSymToIdOverrides[coinData.symbol] = coinData.id.toString()
  })

  return { coinmarketcap: adapterSymToIdOverrides }
}

const createSymbolToIdOverridesFile = async (): Promise<void> => {
  const symbolToIdOverrides = await getSymbolToIdOverrides()
  fs.writeFileSync('symbolToIdOverrides.json', JSON.stringify(symbolToIdOverrides), {
    encoding: 'utf-8',
  })
}

createSymbolToIdOverridesFile()
