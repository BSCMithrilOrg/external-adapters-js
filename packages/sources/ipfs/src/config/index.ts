import { Config } from '@chainlink/ea-bootstrap'
import { Requester } from '@chainlink/ea-bootstrap'

export const NAME = 'IPFS'
export const DEFAULT_ENDPOINT = 'read'
export const DEFAULT_API_URL = 'https://mithrilverse.infura-ipfs.io/ipfs/'

export const makeConfig = (prefix?: string): Config => {
  const config = Requester.getDefaultConfig(prefix)
  config.api.baseURL = config.api.baseURL || DEFAULT_API_URL
  config.defaultEndpoint = DEFAULT_ENDPOINT
  return config
}
