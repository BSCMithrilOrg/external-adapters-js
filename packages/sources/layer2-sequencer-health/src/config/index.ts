import { AdapterConfigError, Requester, util } from '@chainlink/ea-bootstrap'
import { AdapterContext, Config } from '@chainlink/ea-bootstrap'
import { envDefaultOverrides } from './envDefaultOverrides'
import { SequencerProvider } from 'starknet'

export const NAME = 'L2_SEQUENCER_HEALTH'
export const DEFAULT_PRIVATE_KEY =
  '0x0000000000000000000000000000000000000000000000000000000000000001'

export const adapterContext: AdapterContext = { name: NAME, envDefaultOverrides }

export const DEFAULT_ENDPOINT = 'health'

// 2 minutes
export const DEFAULT_DELTA_TIME = 2 * 60 * 1000
// Blocks that replica nodes can fall behind
export const DEFAULT_DELTA_BLOCKS = 6
// milliseconds to consider a timeout transaction (10 secs)
export const DEFAULT_TIMEOUT_LIMIT = 5 * 1000

export enum Networks {
  Arbitrum = 'arbitrum',
  Optimism = 'optimism',
  Metis = 'metis',
  Starkware = 'starkware',
}

export type EVMNetworks = Exclude<Networks, Networks.Starkware>

const DEFAULT_ARBITRUM_RPC_ENDPOINT = 'https://arb1.arbitrum.io/rpc'
const DEFAULT_OPTIMISM_RPC_ENDPOINT = 'https://mainnet.optimism.io'
const DEFAULT_METIS_RPC_ENDPOINT = 'https://andromeda.metis.io/?owner=1088'

export const RPC_ENDPOINTS: Record<EVMNetworks, string | undefined> = {
  [Networks.Arbitrum]: util.getEnv('ARBITRUM_RPC_ENDPOINT') || DEFAULT_ARBITRUM_RPC_ENDPOINT,
  [Networks.Optimism]: util.getEnv('OPTIMISM_RPC_ENDPOINT') || DEFAULT_OPTIMISM_RPC_ENDPOINT,
  [Networks.Metis]: util.getEnv('METIS_RPC_ENDPOINT') || DEFAULT_METIS_RPC_ENDPOINT,
}

const DEFAULT_OPTIMISM_HEALTH_ENDPOINT = 'https://mainnet-sequencer.optimism.io/health'
const DEFAULT_METIS_HEALTH_ENDPOINT = 'https://tokenapi.metis.io/andromeda/health'
export const HEALTH_ENDPOINTS = {
  [Networks.Arbitrum]: {
    endpoint: util.getEnv('ARBITRUM_HEALTH_ENDPOINT'),
    responsePath: [],
  },
  [Networks.Optimism]: {
    endpoint: util.getEnv('OPTIMISM_HEALTH_ENDPOINT') || DEFAULT_OPTIMISM_HEALTH_ENDPOINT,
    responsePath: ['healthy'],
  },
  [Networks.Metis]: {
    endpoint: util.getEnv('METIS_HEALTH_ENDPOINT') || DEFAULT_METIS_HEALTH_ENDPOINT,
    responsePath: ['healthy'],
  },
  [Networks.Starkware]: {
    endpoint: util.getEnv('STARKWARE_HEALTH_ENDPOINT'),
    responsePath: [],
  },
}

export interface ExtendedConfig extends Config {
  delta: number
  deltaBlocks: number
  timeoutLimit: number
  starkwareConfig: {
    provider: SequencerProvider
    argentAccountAddr: string
  }
}

export const makeConfig = (prefix?: string): ExtendedConfig => {
  const isCacheEnabled = util.parseBool(util.getEnv('CACHE_ENABLED', undefined, adapterContext))
  if (isCacheEnabled) {
    throw new AdapterConfigError({ message: 'Cache cannot be enabled on this adapter' })
  }
  const config = Requester.getDefaultConfig(prefix)
  const delta = Number(util.getEnv('DELTA', prefix)) || DEFAULT_DELTA_TIME
  const deltaBlocks = Number(util.getEnv('DELTA_BLOCKS', prefix)) || DEFAULT_DELTA_BLOCKS
  const timeoutLimit = Number(util.getEnv('NETWORK_TIMEOUT_LIMIT', prefix)) || DEFAULT_TIMEOUT_LIMIT
  config.defaultEndpoint = DEFAULT_ENDPOINT
  const starkwareConfig = instantiateStarkwareConfig()

  return { ...config, delta, deltaBlocks, timeoutLimit, starkwareConfig }
}

const DEFAULT_STARKWARE_SEQUENCER_ENDPOINT = 'https://alpha-mainnet.starknet.io'
const DEFAULT_STARKWARE_FEEDER_GATEWAY_URL = 'feeder_gateway'
const DEFAULT_STARKWARE_GATEWAY_URL = 'gateway'
const DEFAULT_STARKWARE_ARGENT_ACCOUNT_ADDR =
  '0x163995f6cbc4e9e3908ce6161a0bef4459847b42077be419e257c7f837a224a'

const instantiateStarkwareConfig = (): ExtendedConfig['starkwareConfig'] => {
  const baseUrl =
    util.getEnv('STARKWARE_SEQUENCER_ENDPOINT') || DEFAULT_STARKWARE_SEQUENCER_ENDPOINT
  const feederGatewayUrl =
    util.getEnv('STARKWARE_FEEDER_GATEWAY_URL') || DEFAULT_STARKWARE_FEEDER_GATEWAY_URL
  const gatewayUrl = util.getEnv('STARKWARE_GATEWAY_URL') || DEFAULT_STARKWARE_GATEWAY_URL
  const argentAccountAddr =
    util.getEnv('STARKWARE_ARGENT_ACCOUNT_ADDR') || DEFAULT_STARKWARE_ARGENT_ACCOUNT_ADDR
  return {
    provider: new SequencerProvider({
      baseUrl: DEFAULT_STARKWARE_SEQUENCER_ENDPOINT,
      feederGatewayUrl: `${baseUrl}/${feederGatewayUrl}`,
      gatewayUrl: `${baseUrl}/${gatewayUrl}`,
    }),
    argentAccountAddr: argentAccountAddr,
  }
}
