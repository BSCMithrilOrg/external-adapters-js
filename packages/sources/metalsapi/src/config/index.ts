import { Requester } from '@chainlink/ea-bootstrap'
import { Config } from '@chainlink/ea-bootstrap'

export const NAME = 'METALSAPI'

export const DEFAULT_BASE_URL = 'https://metals-api.com/api/'
export const DEFAULT_ENDPOINT = 'convert'

export const makeConfig = (prefix = ''): Config => {
  const config = Requester.getDefaultConfig(prefix, true)
  console.log(config)

  if (config.apiKey) config.api.headers = { ...config.api.headers, 'x-api-key': config.apiKey }
  config.api.baseURL = config.api.baseURL || DEFAULT_BASE_URL
  config.defaultEndpoint = DEFAULT_ENDPOINT
  return config
}
