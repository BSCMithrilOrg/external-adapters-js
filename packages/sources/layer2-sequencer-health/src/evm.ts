import { AdapterResponseInvalidError, Logger } from '@chainlink/ea-bootstrap'
import { EVMNetworks, ExtendedConfig } from './config'
import { requestBlockHeight } from './network'

export const checkOptimisticRollupBlockHeight = (
  network: EVMNetworks,
): ((config: ExtendedConfig) => Promise<boolean>) => {
  let lastSeenBlock: { block: number; timestamp: number } = {
    block: 0,
    timestamp: 0,
  }

  const _isPastBlock = (block: number) => block <= lastSeenBlock.block
  const _isStaleBlock = (block: number, delta: number): boolean => {
    return _isPastBlock(block) && Date.now() - lastSeenBlock.timestamp >= delta
  }
  // If the request hit a replica node that fell behind, the block could be previous to the last seen. Including a deltaBlocks range to consider this case.
  const _isValidBlock = (block: number, deltaBlocks: number) =>
    lastSeenBlock.block - block <= deltaBlocks
  const _updateLastSeenBlock = (block: number): void => {
    lastSeenBlock = {
      block,
      timestamp: Date.now(),
    }
  }

  return async (config: ExtendedConfig): Promise<boolean> => {
    const { delta, deltaBlocks } = config
    const block = await requestBlockHeight(network)
    if (!_isValidBlock(block, deltaBlocks))
      throw new AdapterResponseInvalidError({
        message: `Block found #${block} is previous to last seen #${lastSeenBlock.block} with more than ${deltaBlocks} difference`,
      })
    if (!_isStaleBlock(block, delta)) {
      if (!_isPastBlock(block)) _updateLastSeenBlock(block)
      Logger.info(
        `Block #${block} is not considered stale at ${Date.now()}. Last seen block #${
          lastSeenBlock.block
        } was at ${lastSeenBlock.timestamp}`,
      )
      return true
    }
    Logger.warn(
      `Block #${block} is considered stale at ${Date.now()}. Last seen block #${
        lastSeenBlock.block
      } was at ${lastSeenBlock.timestamp}, more than ${delta} milliseconds ago.`,
    )
    return false
  }
}
