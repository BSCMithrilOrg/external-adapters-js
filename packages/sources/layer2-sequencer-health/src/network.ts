import { Logger, Requester } from '@chainlink/ea-bootstrap'
import { HEALTH_ENDPOINTS, Networks, ExtendedConfig } from './config'

import { sendDummyStarkwareTransaction } from './starkware'
import { sendEVMDummyTransaction } from './evm'
import { sleep } from '@chainlink/ea-bootstrap/src/lib/util'

const NO_ISSUE_MSG =
  'This is an error that the EA uses to determine whether or not the L2 Sequencer is healthy.  It does not mean that there is an issue with the EA.'

// These errors come from the Sequencer when submitting an empty transaction
const sequencerOnlineErrors: Record<Networks, string[]> = {
  [Networks.Arbitrum]: ['gas price too low', 'forbidden sender address'],
  // TODO: Optimism error needs to be confirmed by their team
  [Networks.Optimism]: ['cannot accept 0 gas price transaction'],
  [Networks.Metis]: ['cannot accept 0 gas price transaction'],
  [Networks.Starkware]: [
    'StarknetErrorCode.UNINITIALIZED_CONTRACT',
    'StarknetErrorCode.OUT_OF_RANGE_FEE',
  ],
}

export interface NetworkHealthCheck {
  (network: Networks, config: ExtendedConfig): Promise<undefined | boolean>
}

export interface ResponseSchema {
  result: number
}

export const getSequencerHealth: NetworkHealthCheck = async (
  network: Networks,
): Promise<undefined | boolean> => {
  if (!HEALTH_ENDPOINTS[network]?.endpoint) {
    Logger.info(`Health endpoint not available for network: ${network}`)
    return
  }
  const response = await Requester.request({
    url: HEALTH_ENDPOINTS[network]?.endpoint,
  })
  const isHealthy = !!Requester.getResult(response.data, HEALTH_ENDPOINTS[network]?.responsePath)
  Logger.info(
    `Health endpoint for network ${network} returned a ${
      isHealthy ? 'healthy' : 'unhealthy'
    } response`,
  )
  return isHealthy
}

export const getStatusByTransaction = async (
  network: Networks,
  config: ExtendedConfig,
): Promise<boolean> => {
  let isSequencerHealthy = true
  try {
    Logger.info(`Submitting empty transaction for network: ${network}`)
    await sendEmptyTransaction(network, config)
  } catch (e) {
    isSequencerHealthy = isExpectedErrorMessage(network, e as Error)
  }
  return isSequencerHealthy
}

const sendEmptyTransaction = async (network: Networks, config: ExtendedConfig): Promise<void> => {
  switch (network) {
    case Networks.Starkware:
      await sendDummyStarkwareTransaction(config)
      break
    default:
      await sendEVMDummyTransaction(network, config.timeoutLimit)
  }
}

const isExpectedErrorMessage = (network: Networks, e: Error) => {
  const _getErrorMessage = (e: Error): string => {
    const paths = {
      [Networks.Arbitrum]: ['error', 'message'],
      [Networks.Optimism]: ['error', 'message'],
      [Networks.Metis]: ['error', 'message'],
      [Networks.Starkware]: ['errorCode'],
    }
    return (Requester.getResult(e, paths[network]) as string) || ''
  }
  const error = e as Error
  if (sequencerOnlineErrors[network].includes(_getErrorMessage(error))) {
    Logger.debug(`Transaction submission failed with an expected error ${_getErrorMessage(error)}.`)
    return true
  }
  Logger.error(
    `Transaction submission failed with an unexpected error. ${NO_ISSUE_MSG} Error Message: ${error.message}`,
  )
  return false
}

export async function retry<T>({
  promise,
  retryConfig,
}: {
  promise: () => Promise<T>
  retryConfig: ExtendedConfig['retryConfig']
}): Promise<T> {
  let numTries = 0
  let error
  while (numTries < retryConfig.numRetries) {
    try {
      return await promise()
    } catch (e) {
      error = e
      numTries++
      await sleep(retryConfig.retryInterval)
    }
  }
  throw error
}

export function race<T>({
  promise,
  timeout,
  error,
}: {
  promise: Promise<T>
  timeout: number
  error: string
}): Promise<T> {
  let timer: NodeJS.Timeout

  return Promise.race([
    new Promise((_, reject) => {
      timer = setTimeout(reject, timeout, error)
    }) as Promise<T>,
    promise.then((value) => {
      clearTimeout(timer)
      return value
    }),
  ])
}
