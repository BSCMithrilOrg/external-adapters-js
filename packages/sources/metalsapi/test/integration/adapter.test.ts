import { AdapterRequest } from '@chainlink/ea-bootstrap'
import * as process from 'process'
import { server as startServer } from '../../src'
import {
  mockResponseSuccessConvertEndpoint,
  mockResponseSuccessLatestEndpoint,
  mockResponseSuccessLatestBtcEndpoint,
  mockResponseSuccessLatestUsdInverseEndpoint,
} from './fixtures'
import { setupExternalAdapterTest } from '@chainlink/ea-test-helpers'
import type { SuiteContext } from '@chainlink/ea-test-helpers'
import { SuperTest, Test } from 'supertest'

describe('execute', () => {
  const id = '1'
  const context: SuiteContext = {
    req: null,
    server: startServer,
  }

  const envVariables = {
    API_KEY: process.env.API_KEY || 'fake-api-key',
  }

  setupExternalAdapterTest(envVariables, context)
  describe('convert api', () => {
    const data: AdapterRequest = {
      id,
      data: {
        endpoint: 'convert',
        base: 'XAU',
        quote: 'USD',
      },
    }

    it('should return success', async () => {
      mockResponseSuccessConvertEndpoint()

      const response = await (context.req as SuperTest<Test>)
        .post('/')
        .send(data)
        .set('Accept', '*/*')
        .set('Content-Type', 'application/json')
        .expect('Content-Type', /json/)
        .expect(200)
      expect(response.body).toMatchSnapshot()
    })
  })

  describe('latest api', () => {
    it('returns success with single base/quote pair', async () => {
      const data: AdapterRequest = {
        id,
        data: {
          endpoint: 'latest',
          base: 'XAU',
          quote: 'USD',
        },
      }

      mockResponseSuccessLatestEndpoint()

      const response = await (context.req as SuperTest<Test>)
        .post('/')
        .send(data)
        .set('Accept', '*/*')
        .set('Content-Type', 'application/json')
        .expect('Content-Type', /json/)
        .expect(200)
      expect(response.body).toMatchSnapshot()
    })

    it('returns success with batched quote symbols', async () => {
      const data: AdapterRequest = {
        id,
        data: {
          endpoint: 'latest',
          base: 'BTC',
          quote: ['USD', 'XAU'],
        },
      }

      mockResponseSuccessLatestBtcEndpoint()

      const response = await (context.req as SuperTest<Test>)
        .post('/')
        .send(data)
        .set('Accept', '*/*')
        .set('Content-Type', 'application/json')
        .expect('Content-Type', /json/)
        .expect(200)
      expect(response.body).toMatchSnapshot()
    })

    it('should succeed with batched base symbols (inverted quote)', async () => {
      const data: AdapterRequest = {
        id,
        data: {
          endpoint: 'latest',
          base: ['XAU', 'GBP'], //Batching happens in order, so expect "XAU,GBP" to appear in the query (not "GBP,XAU")
          quote: 'USD',
        },
      }
      //The EA should inverse the base and quote. I.e, the above would result in a  {base: USD, quote: [XAU, GBP]
      mockResponseSuccessLatestUsdInverseEndpoint()

      const response = await (context.req as SuperTest<Test>)
        .post('/')
        .send(data)
        .set('Accept', '*/*')
        .set('Content-Type', 'application/json')
        .expect('Content-Type', /json/)
        .expect(200)
      console.log(response.body)
      expect(response.body).toMatchSnapshot()
    })
  })
})
