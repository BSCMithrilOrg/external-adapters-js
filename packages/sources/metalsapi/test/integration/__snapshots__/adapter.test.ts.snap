// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`execute convert api should return success 1`] = `
Object {
  "data": Object {
    "result": 1785.0181286441143,
  },
  "jobRunID": "1",
  "providerStatusCode": 200,
  "result": 1785.0181286441143,
  "statusCode": 200,
}
`;

exports[`execute latest api returns success with batched quote symbols 1`] = `
Object {
  "data": Object {
    "base": "BTC",
    "date": "2022-01-12",
    "rates": Object {
      "USD": 42968.36778447169,
      "XAU": 0.04228229144046888,
    },
    "results": Array [
      Array [
        "2ZgJBx2Rs86dhkQN/BET4ojSkTU=",
        Object {
          "data": Object {
            "base": "BTC",
            "endpoint": "latest",
            "quote": "USD",
          },
          "debug": Object {
            "batchCacheKey": "2E1RbkLbQOXUVtGI1E2+n/etEO8=",
            "batchChildrenCacheKeys": Array [
              Array [
                "2ZgJBx2Rs86dhkQN/BET4ojSkTU=",
                Object {
                  "data": Object {
                    "base": "BTC",
                    "endpoint": "latest",
                    "quote": "USD",
                  },
                  "id": "1",
                },
              ],
            ],
            "cacheKey": "Io9CE5GgK3Zu72UltbQi09+PByY=",
          },
          "id": "1",
          "rateLimitMaxAge": 176470588,
        },
        42968.36778447169,
      ],
      Array [
        "jn7Ay27+0XZwS3+kIquAQibbEtg=",
        Object {
          "data": Object {
            "base": "BTC",
            "endpoint": "latest",
            "quote": "XAU",
          },
          "debug": Object {
            "batchCacheKey": "2E1RbkLbQOXUVtGI1E2+n/etEO8=",
            "batchChildrenCacheKeys": Array [
              Array [
                "2ZgJBx2Rs86dhkQN/BET4ojSkTU=",
                Object {
                  "data": Object {
                    "base": "BTC",
                    "endpoint": "latest",
                    "quote": "USD",
                  },
                  "id": "1",
                },
              ],
            ],
            "cacheKey": "Io9CE5GgK3Zu72UltbQi09+PByY=",
          },
          "id": "1",
          "rateLimitMaxAge": 176470588,
        },
        0.04228229144046888,
      ],
    ],
    "success": true,
    "timestamp": 1641990180,
    "unit": "per ounce",
  },
  "jobRunID": "1",
  "providerStatusCode": 200,
  "statusCode": 200,
}
`;

exports[`execute latest api returns success with single base/quote pair 1`] = `
Object {
  "data": Object {
    "result": 1817.0552439305814,
  },
  "jobRunID": "1",
  "providerStatusCode": 200,
  "result": 1817.0552439305814,
  "statusCode": 200,
}
`;

exports[`execute latest api should succeed with batched base symbols (inverted quote) 1`] = `
Object {
  "data": Object {
    "base": "USD",
    "date": "2022-01-12",
    "rates": Object {
      "GBP": 42968.36778447169,
      "XAU": 0.04228229144046888,
    },
    "results": Array [
      Array [
        "4LsxrIT483BCS93HKnZ32ylFSOU=",
        Object {
          "data": Object {
            "base": "XAU",
            "endpoint": "latest",
            "quote": "USD",
          },
          "debug": Object {
            "batchCacheKey": "aQyZBD4Svg72cpXfGQUCGpzUNQI=",
            "batchChildrenCacheKeys": Array [
              Array [
                "4LsxrIT483BCS93HKnZ32ylFSOU=",
                Object {
                  "data": Object {
                    "base": "XAU",
                    "endpoint": "latest",
                    "quote": "USD",
                  },
                  "id": "1",
                },
              ],
            ],
            "cacheKey": "WsOvkm0wEBg6k+OFH4f0i465GGQ=",
          },
          "id": "1",
          "rateLimitMaxAge": 235294117,
        },
        23.65056305919334,
      ],
      Array [
        "s4Eob4V/uG3W8FVSucnYEe/iTP4=",
        Object {
          "data": Object {
            "base": "GBP",
            "endpoint": "latest",
            "quote": "USD",
          },
          "debug": Object {
            "batchCacheKey": "aQyZBD4Svg72cpXfGQUCGpzUNQI=",
            "batchChildrenCacheKeys": Array [
              Array [
                "4LsxrIT483BCS93HKnZ32ylFSOU=",
                Object {
                  "data": Object {
                    "base": "XAU",
                    "endpoint": "latest",
                    "quote": "USD",
                  },
                  "id": "1",
                },
              ],
            ],
            "cacheKey": "WsOvkm0wEBg6k+OFH4f0i465GGQ=",
          },
          "id": "1",
          "rateLimitMaxAge": 235294117,
        },
        0.00002327293429008,
      ],
    ],
    "success": true,
    "timestamp": 1641990180,
    "unit": "per ounce",
  },
  "jobRunID": "1",
  "providerStatusCode": 200,
  "statusCode": 200,
}
`;
